File: Bubble Sort/bubble_sort.py
# Jacobus Burger (2022)
# Bubble Sort (Python 3)
# Bubble Sort is a straightforwards sorting algorithm and often the first
#   taught in Compting Science courses. It's time complexity of O(n^2)
#   is "bad" on large arrays, but it's simple design makes it
#   straightforward to implement and effective to use on smaller
#   arrays. Overall, it's a decent if inefficient sorting algorithm!
# see:
# - https://en.wikipedia.org/wiki/Bubble_sort
# - https://rosettacode.org/wiki/Sorting_algorithms/Bubble_sort#Python
from random import randint


# Time Complexity: O(n^2)
# Space Complexity: O(1)
def sort(array: list) -> list:
    unsorted = True
    while unsorted:
        unsorted = False
        for i in range(1, len(array)):
            if array[i - 1] > array[i]:
                array[i - 1], array[i] = array[i], array[i - 1]
                unsorted = True


if __name__ == "__main__":
    # get input
    n = int(input())
    array = [randint(0, n) for _ in range(n)]

    # show unsorted
    # NOTE: cool snakelang trick to avoid looping when printing a list
    print(*array, sep=' ')

    # sort
    sort(array)

    # show sorted
    print(*array, sep=' ')

File: Bubble Sort/bubble_sort.lua
--[[
  Jacobus Burger (2025-09-24)
  Bubble Sort (Lua 5.4)
  Bubble Sort is a straightforwards sorting algorithm and often the first
    taught in Compting Science courses. It's time complexity of O(n^2)
    is "bad" on large arrays, but it's simple design makes it
    straightforward to implement and effective to use on smaller
    arrays. Overall, it's a decent if inefficient sorting algorithm!
  see:
  - https://en.wikipedia.org/wiki/Bubble_sort
  - https://rosettacode.org/wiki/Sorting_algorithms/Bubble_sort#Lua
]]


local function sort(array)
  for _ = 1, #array do
    for i = 2, #array do
      if array[i - 1] > array[i] then
        array[i - 1], array[i] = array[i], array[i - 1]
      end
    end
  end
end


local function main()
  -- get input
  local n = io.read("n")
  -- sadly, no comprehensions in moonlang 🌕 like in snakelang 🐍
  math.randomseed(os.time())
  local array = {}
  for _ = 1, n do
    table.insert(array, math.random(n))
  end

  -- show unsorted
  -- moonlang 🌕 does have nice features like table.concat
  print(table.concat(array, ' '))

  -- sort
  sort(array)

  -- show sorted
  print(table.concat(array, ' '))
end


main()
